### PR TITLE
Mongo pooling

### PR DESCRIPTION
**Pull request by italolelis**
## What does this PR do?

- Lots of refactoring on how the server is created and handled (Preparing to have multiple providers)
- Added watch to mongodb, filesystem and in memory providers

### CODE REVIEW

File: pkg/api/mongodb_repository.go [hunk 1]
```diff
@@ -12,12 +16,56 @@ const (
...
+		for {
+			select {
+			case <-refreshTicker.C:
+				defs, err := r.FindAll()
+				if err != nil {
+					log.WithError(err).Error("Failed to get configurations on watch")
+					continue
+				}
+
+				cfgChan <- ConfigurationChanged{
```
Comment (vgarvardt) [inline pkg/api/mongodb_repository.go:92]: Are you calculating diff between current defs list and this one on the router/reciever level or just reload the router with every timer tick?

**UPD** I think I found the answer https://github.com/hellofresh/janus/pull/305/files#diff-09856fe9becddf0199651f451409356aR173

File: pkg/proxy/register.go [hunk 2]
```diff
@@ -28,6 +28,11 @@ func NewRegister(router router.Router, params Params) *Register {
 	return &Register{router, params}
 }
 
+// UpdateRouter updates the reference to the router. This is useful to reload the mutex
```
Comment (vgarvardt) [inline pkg/proxy/register.go]: `reload the mutex` - what is `mutex` in this case? maybe `mux`?

File: pkg/api/mongodb_repository.go (hunk 1 above)
Comment (italolelis) [reply pkg/api/mongodb_repository.go:92]: Yes, I calculate the diff with `reflect.DeepEquals` like `traefik` does

File: cmd/janus/context.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+// ContextWithSignal create a context cancelled when SIGINT or SIGTERM are notified
+func ContextWithSignal(ctx context.Context) context.Context {
+	newCtx, cancel := context.WithCancel(ctx)
+	signals := make(chan os.Signal)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		select {
+		case <-signals:
+			cancel()
```
Comment (rafaeljesus) [inline cmd/janus/context.go:18]: you can also `close(signals)`

File: pkg/api/mongodb_repository.go [hunk 4]
```diff
@@ -12,12 +16,56 @@ const (
...
+	log.Debug("Connected to MongoDB")
+	session.SetMode(mgo.Monotonic, true)
+
+	return &MongoRepository{Session: session, refreshTime: refreshTime}, nil
+}
+
+// Close terminates the session.  It's a runtime error to use a session
+// after it has been closed.
+func (r *MongoRepository) Close() error {
+	r.Session.Close()
```
Comment (rafaeljesus) [inline pkg/api/mongodb_repository.go:42]: shouldn't we `return r.Session.Close()`?

File: pkg/api/repository.go [hunk 5]
```diff
@@ -17,13 +18,15 @@ const (
 
 // Repository defines the behavior of a proxy specs repository
 type Repository interface {
+	io.Closer
```
Comment (rafaeljesus) [inline pkg/api/repository.go:21]: 👍

File: pkg/api/repository.go [hunk 6]
```diff
@@ -17,13 +18,15 @@ const (
...
 	FindAll() ([]*Definition, error)
 	FindByName(name string) (*Definition, error)
 	FindByListenPath(path string) (*Definition, error)
 	Exists(def *Definition) (bool, error)
 	Add(app *Definition) error
 	Remove(name string) error
 	FindValidAPIHealthChecks() ([]*Definition, error)
+	Watch(ctx context.Context, cfgChan chan<- ConfigurationChanged)
```
Comment (rafaeljesus) [inline pkg/api/repository.go]: we could be a interface as well `Watcher`

Comment (rafaeljesus) [reply pkg/api/repository.go]: which exposes `Watch` method

File: pkg/server/server.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+// Server is the Janus server
+type Server struct {
+	server                *http.Server
+	provider              api.Repository
+	register              *proxy.Register
+	defLoader             *loader.APILoader
+	currentConfigurations []*api.Definition
+	configurationChan     chan api.ConfigurationChanged
+	stopChan              chan bool
```
Comment (rafaeljesus) [inline pkg/server/server.go]: ideally it should be a struct

File: pkg/server/server.go [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+	go func(ctx context.Context) {
+		<-ctx.Done()
+		if ctx.Err() == context.Canceled {
+			return
+		} else if ctx.Err() == context.DeadlineExceeded {
+			panic("Timeout while stopping janus, killing instance ✝")
+		}
+	}(ctx)
+
+	close(s.stopChan)
```
Comment (rafaeljesus) [inline pkg/server/server.go]: Usually closing resources stays in the top wrapped by defer e.g:

```go
func (s *Server) Close() error {
  defer close(s.stopChan)
  defer close(s.configurationChan)

  // code goes here
}
```

File: pkg/server/server.go [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+		log.Info("Stopping server gracefully")
+		s.Close()
+	}()
+	go func() {
+		if err := s.startHTTPServers(); err != nil {
+			log.WithError(err).Fatal("Could not start http servers")
+		}
+	}()
+	go func() {
+		s.listenProviders(s.stopChan)
```
Comment (rafaeljesus) [inline pkg/server/server.go]: it could be just `go s.listenProviders(s.stopChan)`

File: pkg/server/server.go [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+		defer s.Close()
+		<-ctx.Done()
+		log.Info("I have to go...")
+		reqAcceptGraceTimeOut := time.Duration(s.globalConfig.GraceTimeOut)
+		if reqAcceptGraceTimeOut > 0 {
+			log.Infof("Waiting %s for incoming requests to cease", reqAcceptGraceTimeOut)
+			time.Sleep(reqAcceptGraceTimeOut)
+		}
+		log.Info("Stopping server gracefully")
+		s.Close()
```
Comment (rafaeljesus) [inline pkg/server/server.go]: extra call of Close method

File: pkg/api/mongodb_repository.go (hunk 4 above)
Comment (italolelis) [reply pkg/api/mongodb_repository.go:42]: `r.Session.Close()` from mongo returns void, so I can't return it

**Review by rafaeljesus — APPROVED** [review]

**Pull request merged**